### PR TITLE
[SPIKE] Optional types API

### PR DESCRIPTION
**Pull request by itsdouges**
TBD

### CODE REVIEW

File: babel.config.json [hunk 1]
```diff
@@ -19,7 +19,10 @@
           {
             "nonce": "\"k0Mp1lEd\"",
             "importReact": false,
-            "optimizeCss": false
+            "optimizeCss": false,
+            "customModuleOrigins": [
```
Comment (itsdouges) [inline babel.config.json:23]: What setting the config looks like for new modules

File: packages/babel-plugin/src/babel-plugin.ts [hunk 2]
```diff
@@ -80,7 +86,9 @@ export default declare<State>((api) => {
               const jsxSourceMatches = JSX_SOURCE_ANNOTATION_REGEX.exec(comment.value);
               const jsxMatches = JSX_ANNOTATION_REGEX.exec(comment.value);
 
-              if (jsxSourceMatches && jsxSourceMatches[1] === COMPILED_MODULE) {
+              // jsxPragmas currently only run on the top-level compiled module,
+              // hence we don't interrogate this.compiledModuleOrigins.
+              if (jsxSourceMatches && jsxSourceMatches[1] === DEFAULT_COMPILED_MODULE) {
```
Comment (itsdouges) [inline packages/babel-plugin/src/babel-plugin.ts:91]: Call out: jsx pragma doesn't support custom modules. I didn't think it was worth the effort as it's annoying to set up as a consumer.

File: packages/babel-plugin/src/babel-plugin.ts [hunk 3]
```diff
@@ -159,7 +167,27 @@ export default declare<State>((api) => {
...
       ImportDeclaration(path, state) {
-        if (path.node.source.value !== COMPILED_MODULE) {
+        const userLandModule = path.node.source.value;
+
+        const isCompiledModule = this.compiledModuleOrigins.some((compiledModuleOrigin) => {
+          if (userLandModule === DEFAULT_COMPILED_MODULE) {
+            return true;
+          }
+
+          if (
```
Comment (itsdouges) [inline packages/babel-plugin/src/babel-plugin.ts:177]: TODO: Add a unit test for the babel plugin. I just tested it as an integration test in the react package, but best to test out edge cases. Was a bit naughty.

File: packages/react/src/create-strict-api/__tests__/__fixtures__/strict-api.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import { createStrictAPI } from '../..';
+
+const { css, XCSSProp, cssMap, cx } = createStrictAPI<{
+  '&:hover': {
+    color: 'var(--ds-text)';
+    background: 'var(--ds-surface-hover)' | 'var(--ds-surface-sunken-hover)';
+  };
+  background: 'var(--ds-surface)' | 'var(--ds-surface-sunken';
+  bkgrnd: 'red' | 'green';
```
Comment (itsdouges) [inline packages/react/src/create-strict-api/__tests__/__fixtures__/strict-api.ts:9]: Currently passing unknown values isn't a type violation. I want to spend some time to see if it could be.

File: packages/react/src/create-strict-api/__tests__/index.test.tsx [hunk 5]
```diff
@@ -0,0 +1,310 @@
+/** @jsxImportSource @compiled/react */
```
Comment (itsdouges) [inline packages/react/src/create-strict-api/__tests__/index.test.tsx:1]: Try to take close care when looking at this file, look at the type violations and see if it should (or shouldn't!) be one.

File: packages/react/src/types.ts [hunk 6]
```diff
@@ -96,8 +96,13 @@ export type CSSPseudos =
...
+ * if it takes a subset of them. This ensures the (lack-of an)
  * excess property check doesn't enable makers to circumvent the
  * system and pass in values they shouldn't.
  */
 export type CSSProperties = Readonly<CSS.Properties<string | number>>;
+
+/**
+ * Words go here brrr
+ */
+export type CSSPropertiesStrict = Readonly<CSS.StandardProperties & CSS.SvgProperties>;
```
Comment (itsdouges) [inline packages/react/src/types.ts:108]: This type removes all of the "MozXyz" properties.

Comment (MonicaOlejniczak) [reply packages/react/src/types.ts:108]: Prefer the naming as `StrictCSSProperties`

File: packages/react/src/create-strict-api/__tests__/index.test.tsx [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+
+    it('should enforce required properties', () => {
+      function Button({
+        xcss,
+      }: {
+        xcss: ReturnType<
+          typeof XCSSProp<
+            'background',
+            never,
+            { requiredProperties: 'background'; requiredPseudos: never }
```
Comment (MonicaOlejniczak) [inline packages/react/src/create-strict-api/__tests__/index.test.tsx:287]: I find this argument a bit strange, can't consumers construct it instead using combinations of `Partial` / `Required` etc?

In this model, would it be possible to do something like this? I haven't looked at the return type too closely

```
// assume default is partial, but it's based on the schema really
ReturnType<typeof XCSSProp<'background', never>>
  & Required<ReturnType<typeof XCSSProp<'color', never>>>
```

File: babel.config.json (hunk 1 above)
Comment (itsdouges) [reply babel.config.json:23]: Yeah all new names need some bikeshedding. Import sources could work — we have an option to explicitly tie the name of the property to the strict api stuff or we can keep it generic.

File: packages/babel-plugin/src/babel-plugin.ts [hunk 8]
```diff
@@ -59,6 +59,12 @@ export default declare<State>((api) => {
       this.pathsToCleanup = [];
       this.pragma = {};
       this.usesXcss = false;
+      this.compiledModuleOrigins = [
+        DEFAULT_COMPILED_MODULE,
+        ...(this.opts.customModuleOrigins
+          ? this.opts.customModuleOrigins.map((origin) => join(process.cwd(), origin))
```
[the remark this answers is not in the document]
Comment (itsdouges) [reply packages/babel-plugin/src/babel-plugin.ts:65]: Agreed, this didn't feel nice when I did it. It being relative to the babel config is sensible, do we have the config path during a pass? I'll look.

File: packages/react/src/create-strict-api/__tests__/index.test.tsx (hunk 7 above)
Comment (itsdouges) [reply packages/react/src/create-strict-api/__tests__/index.test.tsx:287]: We technically could, but this way allows it to be:

1. type safe, it has to be a value passed into the generic 1st/2nd args
2. easier, less of a headache to type out / ensure it's correct

I'm for keeping it as is.

File: packages/babel-plugin/src/babel-plugin.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (itsdouges) [reply packages/babel-plugin/src/babel-plugin.ts:65]: Ah, but we might also not always have a Babel config. Could be Webpack or Parcel... let me think.

File: packages/react/src/create-strict-api/__tests__/__fixtures__/strict-api.ts (hunk 4 above)
Comment (itsdouges) [reply packages/react/src/create-strict-api/__tests__/__fixtures__/strict-api.ts:10]: I looked at this: The excess property check is killing us. Going to leave it — if anyone has ideas LMK!

File: packages/react/src/create-strict-api/index.ts [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+   * <div css={styles.solid} />
+   * ```
+   */
+  cssMap<
+    TStyles extends Record<
+      string,
+      StrictCSSProperties & PseudosDeclarations & EnforceSchema<TSchema>
+    >
+  >(
+    styles: TStyles
```
Comment (kylorhall-atlassian) [inline packages/react/src/create-strict-api/index.ts:62]: @itsdouges I know this will feel redundant, and it is, but I think this will fix the issue (give or take), just copy the `extends` above:
```diff
-    styles: TStyles
+    styles: Record<
+       string,
+       StrictCSSProperties & PseudosDeclarations & EnforceSchema<TSchema>
+     >
```

Currently this is basically just `TStyles` must extend this, which seems right, but technically this **_extends_** `{ color: 'var(--ds-text)' }`:
```
cssMap({
  default: {
    color: 'var(--ds-text)',
    '* > button': { color: 'transparent !important' },
  },
});
```

`extends` is an annoying keyword, I wish there was a `TStyles satisfies …` in generics, but I keep going in circles in TS issues ([like this one](https://github.com/microsoft/TypeScript/issues/51679)) and can't figure out why they don't want to make this change…

Before: [TS Playground(https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBDAnmApnAVHAhgZzgYQGVC4AzKCEOAcgGMcclVqBuAKDZQA9JYFk0RQgAUcKAK4ATCHgC8bOHAA+NAGQAudVlIwUUaguVrNAIyy0A1pIpgDildQ3qTKUtBR2jjzbXEfDDk6+KAC0UCgA5sAQAHaegZqkwFCMIQA2KDC6+gHG6kkpMOnAMf72eRFQWCAgWFAhehQ55d7qtVAWevF5YGnmKAAWEGmSXbmtYhm0MNFx4044qGlpJRENUE3drTB1EZkhulwwW04AbsAoAO4HVTE4wDOxJ1rTwKdlXhpY4jAQScvPEx9GIWZ60AYoSwoSTPUakb5pY7zdRwkrQ2HAHBYIHo5EocBIZ4oGLYjIw5EFRjPNy+HDUiC0kLne44+mMy4PAYlaniZY4WjhYnPIbvZqfdQlMJYGJ7Z4lUbZEAlLC6OUxU5YFbkloaDI6Z4rEEGhma4pG5EQMCPElpZ4QH4hP5SmUfBJgb5ibXisDAabicIhSU+v3hZ69fpDEZ6EI4IaXOY69ThxCrZ7hLCSR0xNKINMoDMhS5QB6u4zhACO4mSuMTxYiAyRiZ2UD2jfF4jE9RKGq1zw70Z7wC9CUHw+MzJLMPYHG4vHgTEExGENj0MxQcjgACV89Js4gADxCAB0y8tq4uOH3jGLMqMMXEIBcUAAfM-pwuCJQfWTCOD8Vg4FkOBCBgYtpiEU9UFgC84FUOBRAkaQcAAEUhPoqmtHBp1naB5wEYDQN9GAIJXaD10Arcd1iHND2II8QOlSQ6kkSDz3IuDj0IU4IlYsicDfDgPwQqQZFQ2h0JVWYNwAb0MABtABFOASgIJcxBEnAAF0AH51AIsDiKXUi1ywtgAF933wgBRGI3CgWgUF-CFan3AAVAB5EwACtIRgZ8KNkxQ5OEZSYjgTpED+OAPO83ydL0kLuF0GJJDwCKovwL9gB-P9akMRRtOizyfOmYLNLgJLiVSyjaGgSQr0ImUABo4HEEEYggeNn3yxQ4EKmy7Icpz-zc4rfLK7ret6vSYpKmAyp6vTSlFdgLME-DMvAbLoV4pAxIk61t1IPRiQcgL5OoAB9S78HcgBZYQAEkABkrOQy7hE3dzhCszdXIATUu5CrPwZ6AEFNzB1zHvcgA5S7NysgAxX6rNh-ArKB9zLth9zXMugB1TdHtczHkMexH8Fc57Acu6hNL00C-FW7CeFw-hUE-LafyQDJL1m3z-KAwK4EU0LwpQSLSCK2LpgZmW5sU8rKpSvBt1qqB6uvVYWraiwOq6nrCs279oRAxA+dG2X5oUzTJumrnTZY4zEH2upJNiI6TpiByWY4Wq7ngAV810Ci3OG2oKqOKq8BN7bJAjrBnwACgASkA-yRcUUg2teWI4HoHA7qwMA4H3Hqptc82+aj5LqvVuqGpvCIWpAgySLPPjYPg9SkLdjCpO7gboCG3KsHDsfXx6lOK96xgLfXGbq-XHrU70rOprgdNdxzUWQpU9Lpar3n13luOeYX-nl5wCb2Cmsy4A3qaAHpn7gAABGAcAaHhfPWJo4AhBCHAS4AwVQoFFLPRQ4QYD+jCvPPmd9epmQ4FNGBcDH4FwYMXUuKCUH+1iIwTBhccFwAfkBYO4D9wb1qmkaAelqAaigMnIBqUDhJVTtQJqPVvAii6OvAuwx6E0CYSwkIbDDhFD4VAThZDzIp1TtOAORCEEoFIRQ7BJdk4bw6qUdeUDBF0KgAw0RrDv6SM4dwzevVeEQFFNQARtDhGMLqGIiRSUQjSNkWZKx1jjAdRgMnfIyRUjgmypIdQASgmUiKGEkYqdOGOKEcYmg4QYRkN8dYxwcB-ImB+L8OISSjEMKBH4agcj77mVTkAA)
After: [TS Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBDAnmApnAVHAhgZzgYQGVC4AzKCEOAcgGMcclVqBuAKDZQA9JYFk0RQgAUcKAK4ATCHgC8bOHAA+NAGQAudVlIwUUaguVrNAIyy0A1pIpgDildQ3qTKUtBR2jjzbXEfDDk6+KAC0UCgA5sAQAHaegZqkwFCMIQA2KDC6+gHG6kkpMOnAMf72eRFQWCAgWFAhehQ55d7qtVAWevF5YGnmKAAWEGmSXbmtYhm0MNFx4044qGlpJRENUE3drTB1EZkhulwwW04AbsAoAO4HVTE4wDOxJ1rTwKdlXhpY4jAQScvPEx9GIWZ60AYoSwoSTPUakb5pY7zdRwkrQ2HAHBYIHo5EocBIZ4oGLYjIw5EFRjPNy+HDUiC0kLne44+mMy4PAYlaniZY4WjhYnPIbvZqfdQlMJYGJ7Z4lUbZEAlLC6OUxU5YFbkloaDI6Z4rEEGhma4pG5EQMCPElpZ4QH4hP5SmUfBJgb5ibXisDAabicIhSU+v3hZ69fpDEZ6EI4IaXOY69ThxCrZ7hLCSR0xNKINMoDMhS5QB6u4zhACO4mSuMTxYiAyRiZ2UD2jfF4jE9RKGq1zw70Z7wC9CUHw+MzJLMPYHG4vHgTEExGENj0MxQcjgACV89Js4gADxCAB0y8tq4uOH3jGLMqMMXEIBcUAAfM-pwuCJQfWTCOD8Vg4FkOBCBgYtpiEU9UFgC84FUOBRAkaQcAAEUhPoqmtHBp1naB5wEYDQN9GAIJXaD10Arcd1iHND2II8QOlSQ6kkSDz3IuDj0IU4IlYsicDfDgPwQqQZFQ2h0JVWYNwAb0MABtABFOASgIJcxBEnAAF0AH51AIsDiKXUi1ywtgAF933wgBRGI3CgWgUF-CFan3AAVAB5EwACtIRgZ8KNkxQ5OEZSYjgTpED+OAPO83ydL0kLuF0GJJDwCKovwL9gB-P9akMRRtOizyfOmYLNLgJLiVSyjaGgSQr0ImUABo4HEEEYggeNn3yxQ4EKmy7Icpz-zc4rfLK7ret6vSYpKmAyp6vTSlFdgLME-DMvAbLoV4pAxIk61t1IPRiQcgL5OoAB9S78HcgBZYQAEkABkrOQy7hE3dzhCszdXIATUu5CrPwZ6AEFNzB1zHvcgA5S7NysgAxX6rNh-ArKB9zLth9zXMugB1TdHtczHkMexH8Fc57Acu6hNL00C-FW7CeFw-hUE-LafyQDJL1m3z-KAwK4EU0LwpQSLSCK2LpgZmW5sU8rKpSvBt1qqB6uvVYWraiwOq6nrCs279oRAxA+dG2X5oUzTJumrnTZY4zEH2upJNiI6TpiByWY4Wq7ngAV810Ci3OG2oKqOKq8BN7bJAjrBnwACgASkA-yRcUUg2teWI4HoHA7qwMA4H3Hqptc82+aj5LqvVuqGpvCIWpAgySLPPjYPg9SkLdjCpO7gboCG3KsHDsfXx6lOK96xgLfXPSG81pudf0oiO6gkzu+Evu0PdzCh9skfHLHw8spy5yk-txRU70rOprgdNdxzUWQpU9Lpar3n13luOeYXvzau64JrsCmmZOAD8poAHpoFwAAAIwBwA0Hgvl1hNDgCEEIcBLgDBVCgUUs9FDhBgP6MK88+ZgN6mZDgU0SFkMgQXBgxdS40Jof7WIjBGGFxYXACBQFg74P3A-WqaRoB6WoBqKAycsGpQOElVO1Amo9W8CKLo98C7DHETQKRMiQhyMOEUNRUBFF8PMinVO04A5cIoSgXhAjmEl2Tg-DqpR75EM0WIqAEjdGyOQYYxRyjH69VURAUU1ANGiO0ZIuoeiDFJRCMY0xZkgnBOMB1GAyd8jJFSOCbKkh1AZKyZSIoeSRip0UZErR3iaDhBhHw1JwTHBwH8iYH4vw4hVK8RIoEfhqBmPAeZVOQA)

Comment (itsdouges) [reply packages/react/src/create-strict-api/index.ts:62]: Thanks will look soon. I'll say it's supposed to be a feature to block unknown properties (including pseudos) atm 🤔

Comment (kylorhall-atlassian) [reply packages/react/src/create-strict-api/index.ts:62]: Today it doesn't block them though if your object extends known properties.  It will block if you don't have `color: …` in the above, but once you add it, it won't block.

Before/After playgrounds should be clear…I think I've got this right, I didn't get it all hooked up locally (I don't have this repo setup really).

Comment (itsdouges) [reply packages/react/src/create-strict-api/index.ts:62]: Thanks for the contribution Kylor https://github.com/atlassian-labs/compiled/pull/1582
